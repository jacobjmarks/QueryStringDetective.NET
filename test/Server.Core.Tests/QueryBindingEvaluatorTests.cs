namespace Server.Core.Tests;

public class QueryBindingEvaluatorTests
{
    public static readonly TheoryData<string> TestCases = new()
    {
        "",
        "q=",
        "q=c",
        "q=foo",
        "q=a&q=b&q=c",
        "q=null",
        "q=0",
        "q=1",
        "q=-1",
        "q=true",
        "q=false",
        "q=yes",
        "q=no",
        "q=1&q=2&q=3",
        "q=1&q=2&q=.3",
        "q=-1&q=0&q=1",
        $"q={sbyte.MinValue}",
        $"q={sbyte.MaxValue}",
        $"q={byte.MaxValue}",
        $"q={short.MinValue}",
        $"q={short.MaxValue}",
        $"q={ushort.MaxValue}",
        $"q={int.MinValue}",
        $"q={int.MaxValue}",
        $"q={long.MinValue}",
        $"q={long.MaxValue}",
        $"q={ulong.MaxValue}",
        $"q={Uri.EscapeDataString($"{double.MinValue}")}",
        $"q={Uri.EscapeDataString($"{double.MaxValue}")}",
        $"q={decimal.MinValue}",
        $"q={decimal.MaxValue}",
        $"q={Uri.EscapeDataString($"{float.MinValue}")}",
        $"q={Uri.EscapeDataString($"{float.MaxValue}")}",
        "q=2025-01-13T15:25:00",
        "q=2025-01-13T05:25:00Z",
        "q=2025-01-13",
        "q=15:25:00",
    };

    [Theory]
    [MemberData(nameof(TestCases))]
    public async Task AssertBinding(string queryString)
    {
        using var bindingEvaluator = new QueryBindingEvaluator();
        var results = await bindingEvaluator.EvaluateAsync("?" + queryString);

        var content = $"Input: {queryString}"
            + "\n\nBindings:\n";

        foreach (var result in results)
        {
            content += "\n" + result.Type;
            content += "\n  " + string.Join("\n  ", result.Results
                .Select(r => $"{r.Key}\n    {(
                    r.Value.IsErroneous
                    ? $"e = {r.Value.Error.Message} ({r.Value.Error.Detail})"
                    : $"r = {r.Value.Result}")}"));

            content += "\n";
        }

        // "mock" the current date
        content = content.Replace(DateTime.Now.ToString("yyyy-MM-dd"), "2020-03-12");

        content.Should().MatchSnapshot(SnapshotNameExtension.Create(ToValidFileName(queryString)));
    }

    private static string ToValidFileName(string value)
    {
        foreach (var invalidChar in Path.GetInvalidFileNameChars())
            value = value.Replace(invalidChar, '_');

        return value;
    }
}