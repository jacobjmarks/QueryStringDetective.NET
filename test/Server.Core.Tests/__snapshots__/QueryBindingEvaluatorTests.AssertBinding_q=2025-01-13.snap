Input: q=2025-01-13

Bindings:

string
  MinimalApis
    r = "2025-01-13"
  Controllers
    r = "2025-01-13"

string?
  MinimalApis
    r = "2025-01-13"
  Controllers
    r = "2025-01-13"

string[]
  MinimalApis
    r = ["2025-01-13"]
  Controllers
    r = ["2025-01-13"]

char
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "char q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

char?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<char> q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

char[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "char[] q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

int
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "int q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

int?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<int> q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

int[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "int[] q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

double
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "double q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

double?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<double> q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

double[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "double[] q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

float
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "float q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

float?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<float> q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

float[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "float[] q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

decimal
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "decimal q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

decimal?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<decimal> q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

decimal[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "decimal[] q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

bool
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "bool q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

bool?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<bool> q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

bool[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "bool[] q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

uint
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "uint q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

uint?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<uint> q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

uint[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "uint[] q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

long
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "long q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

long?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<long> q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

long[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "long[] q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

ulong
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ulong q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

ulong?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<ulong> q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

ulong[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ulong[] q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

sbyte
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "sbyte q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

sbyte?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<sbyte> q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

sbyte[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "sbyte[] q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

short
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "short q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

short?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<short> q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

short[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "short[] q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

ushort
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ushort q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

ushort?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<ushort> q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

ushort[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ushort[] q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

byte
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "byte q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

byte?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<byte> q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

byte[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "byte[] q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

DateTime
  MinimalApis
    r = "2025-01-13T00:00:00"
  Controllers
    r = "2025-01-13T00:00:00"

DateTime?
  MinimalApis
    r = "2025-01-13T00:00:00"
  Controllers
    r = "2025-01-13T00:00:00"

DateTime[]
  MinimalApis
    r = ["2025-01-13T00:00:00"]
  Controllers
    r = ["2025-01-13T00:00:00"]

DateTimeOffset
  MinimalApis
    r = "2025-01-13T00:00:00+00:00"
  Controllers
    r = "2025-01-13T00:00:00+10:00"

DateTimeOffset?
  MinimalApis
    r = "2025-01-13T00:00:00+00:00"
  Controllers
    r = "2025-01-13T00:00:00+10:00"

DateTimeOffset[]
  MinimalApis
    r = ["2025-01-13T00:00:00+00:00"]
  Controllers
    r = ["2025-01-13T00:00:00+10:00"]

DateOnly
  MinimalApis
    r = "2025-01-13"
  Controllers
    r = "2025-01-13"

DateOnly?
  MinimalApis
    r = "2025-01-13"
  Controllers
    r = "2025-01-13"

DateOnly[]
  MinimalApis
    r = ["2025-01-13"]
  Controllers
    r = ["2025-01-13"]

TimeOnly
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "TimeOnly q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

TimeOnly?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<TimeOnly> q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)

TimeOnly[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "TimeOnly[] q" from "2025-01-13".)
  Controllers
    e = 400 Bad Request (The value '2025-01-13' is not valid.)
